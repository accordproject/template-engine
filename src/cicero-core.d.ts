declare module '@accordproject/cicero-core';