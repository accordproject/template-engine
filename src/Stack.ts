/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
interface IStack<T> {
    push(item: T): void;
    pop(): T | undefined;
    peek(): T | undefined;
    size(): number;
  }

class Stack<T> implements IStack<T> {
    private storage: T[] = [];

    constructor(private capacity: number = Infinity) {}

    push(item: T): void {
        if (this.size() === this.capacity) {
            throw Error('Stack has reached max capacity, you cannot add more items');
        }
        this.storage.push(item);
    }

    pop(): T | undefined {
        return this.storage.pop();
    }

    peek(): T | undefined {
        return this.storage[this.size() - 1];
    }

    size(): number {
        return this.storage.length;
    }
}