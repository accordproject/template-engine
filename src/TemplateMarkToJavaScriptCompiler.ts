/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import traverse from 'traverse';
import { TemplateMarkModel } from '@accordproject/markdown-common';
import { ClassDeclaration, ModelManager } from '@accordproject/concerto-core';
import { TypeScriptToJavaScriptCompiler } from './TypeScriptToJavaScriptCompiler';
import { TwoSlashReturn } from '@typescript/twoslash';
import { getTemplateClassDeclaration } from './Common';
import { TemplateMarkToTypeScriptCompiler } from './TemplateMarkToTypeScriptCompiler';

const CODE_NODES = [
    `${TemplateMarkModel.NAMESPACE}.FormulaDefinition`,
    `${TemplateMarkModel.NAMESPACE}.ConditionDefinition`,
    `${TemplateMarkModel.NAMESPACE}.ClauseDefinition`,
];

export type CompilerError = {
    nodeId: string;
    code: string;
    errors:TwoSlashReturn['errors'];
};

/**
 * Compiles all the Typescript nodes in a TemplateMark JSON
 * to ES_2020 and returns a modified TemplateMark JSON.
 */
export class TemplateMarkToJavaScriptCompiler {
    modelManager:ModelManager;
    compiler: TypeScriptToJavaScriptCompiler;
    templateClass: ClassDeclaration;

    constructor(modelManager: ModelManager) {
        this.modelManager = modelManager;
        this.compiler = new TypeScriptToJavaScriptCompiler(modelManager);
        this.templateClass = getTemplateClassDeclaration(modelManager);
    }

    compile(templateJson: any) : any {
        const namedTemplateMark = TemplateMarkToTypeScriptCompiler.nameUserCode(templateJson);
        const functionCompiler = new TemplateMarkToTypeScriptCompiler(this.modelManager);

        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const that = this;
        const errors = Array<CompilerError>();
        const compiled = traverse(namedTemplateMark).map(function (x) {
            if (x && CODE_NODES.includes(x.$class)) {
                if (x.code) {  // formula
                    const result = that.compiler.compile(functionCompiler.writeFunctionToString(x.name, 'any', x.code.contents));
                    if(result.errors.length === 0) {
                        x.code.contents = result.code;
                        x.code.type = 'ES_2020';
                        this.update(x);
                    }
                    else {
                        errors.push({
                            nodeId: x.name,
                            code: x.code.contents,
                            errors: result.errors
                        });
                    }
                }
                else if (x.condition) {  // condition or clause (boolean condition)
                    const result = that.compiler.compile(functionCompiler.writeFunctionToString(x.functionName, 'boolean', x.condition.contents));
                    if(result.errors.length === 0) {
                        x.condition.contents = result.code;
                        x.condition.type = 'ES_2020';
                        this.update(x);
                    }
                    else {
                        errors.push({
                            nodeId: x.functionName,
                            code: x.condition.contents,
                            errors: result.errors
                        });
                    }
                }
            }
        });

        if(errors.length === 0) {
            return compiled;
        }
        else {
            throw errors;
        }
    }
}