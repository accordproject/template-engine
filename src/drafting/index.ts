/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import booleanDrafter from './Boolean';
import dateTimeDrafter from './DateTime';
import doubleDrafter from './Double';
import integerDrafter from './Integer';
import longDrafter from './Long';
import monetaryAmountDrafter from './MonetaryAmount';
import stringDrafter from './String';

export function getDrafter(typeName: string) : ((value:any, format:string) => string)|null  {
    switch(typeName) {
    case 'Boolean': return booleanDrafter;
    case 'DateTime': return dateTimeDrafter;
    case 'Double': return doubleDrafter;
    case 'Integer': return integerDrafter;
    case 'Long': return longDrafter;
    case 'org.accordproject.money@0.3.0.MonetaryAmount': return monetaryAmountDrafter;
    case 'String': return stringDrafter;
    default: return null;
    }
}