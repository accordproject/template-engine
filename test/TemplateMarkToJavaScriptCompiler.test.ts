import { ModelManager } from '@accordproject/concerto-core';
import { TemplateMarkTransformer } from '@accordproject/markdown-template';
import { readFileSync } from 'fs-extra';
import { TemplateMarkToJavaScriptCompiler } from '../src/TemplateMarkToJavaScriptCompiler';

describe('templatemark to javascript compiler', () => {
    test('should compile templatemark containing typescript to javascript', async () => {

        const modelManager = new ModelManager();
        modelManager.addCTOModel( readFileSync('./test/templates/good/full/model.cto', 'utf-8'), 'model.cto');
        const compiler = new TemplateMarkToJavaScriptCompiler(modelManager);

        const templateMd = readFileSync('./test/templates/good/full/template.md', 'utf-8');
        const templateMarkTransformer = new TemplateMarkTransformer();
        const templateMarkJson = templateMarkTransformer.fromMarkdownTemplate({ content: templateMd }, modelManager, 'contract', { verbose: false });

        const results = compiler.compile(templateMarkJson);
        expect(results).toMatchSnapshot();
    });
});