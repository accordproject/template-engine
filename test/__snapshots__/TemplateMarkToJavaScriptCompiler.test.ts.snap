// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templatemark to javascript compiler should compile templatemark containing typescript to javascript 1`] = `
Object {
  "$class": "org.accordproject.commonmark@0.5.0.Document",
  "nodes": Array [
    Object {
      "$class": "org.accordproject.templatemark@0.4.0.ContractDefinition",
      "elementType": "test@1.0.0.TemplateData",
      "name": "top",
      "nodes": Array [
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "1",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Welcome!",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Hello ",
            },
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
              "elementType": "String",
              "name": "firstName",
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": " ",
            },
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.JoinDefinition",
              "name": "middleNames",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                  "elementType": "String",
                  "name": "this",
                },
              ],
              "separator": "-",
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": " ",
            },
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.ConditionalDefinition",
              "condition": Object {
                "$class": "org.accordproject.templatemark@0.4.0.Code",
                "contents": "return lastName.startsWith('S')",
                "type": "ES_2020",
              },
              "dependencies": Array [
                "lastName",
              ],
              "functionName": "condition_nodes_0_nodes_1_nodes_5",
              "name": "lastName",
              "whenFalse": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Dude",
                },
              ],
              "whenTrue": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Mister",
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "!",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Middle Names",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.templatemark@0.4.0.ListBlockDefinition",
          "delimiter": "period",
          "name": "middleNames",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.List",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                          "elementType": "String",
                          "name": "this",
                        },
                      ],
                    },
                  ],
                },
              ],
              "tight": "true",
              "type": "bullet",
            },
          ],
          "start": "1",
          "tight": "true",
          "type": "ordered",
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Address",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.templatemark@0.4.0.ClauseDefinition",
          "elementType": "test@1.0.0.Address",
          "name": "address",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.List",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Text",
                          "text": "Street: ",
                        },
                        Object {
                          "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                          "elementType": "String",
                          "name": "street",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Text",
                          "text": "City: ",
                        },
                        Object {
                          "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                          "elementType": "String",
                          "name": "city",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Text",
                          "text": "ZIP: ",
                        },
                        Object {
                          "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                          "elementType": "String",
                          "name": "zip",
                        },
                      ],
                    },
                  ],
                },
              ],
              "tight": "true",
              "type": "bullet",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Your city is: ",
            },
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.WithDefinition",
              "elementType": "test@1.0.0.Address",
              "name": "address",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Emph",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Strong",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                          "elementType": "String",
                          "name": "city",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": ".",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Account Status",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.ConditionalDefinition",
              "name": "active",
              "whenFalse": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your account has been deactivated.",
                },
              ],
              "whenTrue": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your account is active.",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Thank you for visiting us ",
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Strong",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.templatemark@0.4.0.FormulaDefinition",
                      "code": Object {
                        "$class": "org.accordproject.templatemark@0.4.0.Code",
                        "contents": "export function formula_53113a901ca88208df47bc83374866e8d497d84099c0f88123c918ff1960b17e(data, library) {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    const now = dayjs();
    return now.diff(lastVisit, 'day');
}
",
                        "type": "ES_2020",
                      },
                      "dependencies": Array [
                        "now",
                        "lastVisit",
                      ],
                      "name": "formula_53113a901ca88208df47bc83374866e8d497d84099c0f88123c918ff1960b17e",
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": " days ago.",
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Linebreak",
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your last visit was: ",
                },
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.FormattedVariableDefinition",
                  "elementType": "DateTime",
                  "format": "MM/DD/YYYY",
                  "name": "lastVisit",
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": ".",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Orders",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.templatemark@0.4.0.ListBlockDefinition",
          "delimiter": "period",
          "elementType": "test@1.0.0.Order",
          "name": "orders",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.List",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Strong",
                          "nodes": Array [
                            Object {
                              "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                              "elementType": "String",
                              "name": "sku",
                            },
                          ],
                        },
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Text",
                          "text": " : ",
                        },
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Emph",
                          "nodes": Array [
                            Object {
                              "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                              "elementType": "Integer",
                              "name": "amount",
                            },
                          ],
                        },
                      ],
                    },
                  ],
                },
              ],
              "tight": "true",
              "type": "bullet",
            },
          ],
          "start": "1",
          "tight": "true",
          "type": "ordered",
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.OptionalDefinition",
              "elementType": "test@1.0.0.LoyaltyStatus",
              "name": "loyaltyStatus",
              "whenNone": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "You do not have a loyalty status.",
                },
              ],
              "whenSome": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your loyalty status: ",
                },
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                  "elementType": "String",
                  "name": "level",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.templatemark@0.4.0.ClauseDefinition",
          "condition": Object {
            "$class": "org.accordproject.templatemark@0.4.0.Code",
            "contents": "export function condition_nodes_0_nodes_13(data, library) {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    const now = dayjs();
    return preferences.favoriteColors !== undefined && preferences.favoriteColors.length > 0;
}
",
            "type": "ES_2020",
          },
          "elementType": "test@1.0.0.Preferences",
          "functionName": "condition_nodes_0_nodes_13",
          "name": "preferences",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Heading",
              "level": "2",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Favorite Colors",
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your favorite colors are: ",
                },
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.FormulaDefinition",
                  "code": Object {
                    "$class": "org.accordproject.templatemark@0.4.0.Code",
                    "contents": "export function formula_4b8f7e95470eda90057c4648aac4e4c7abb3f93559ed348246b6a15ec1fea473(data, library) {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    const now = dayjs();
    return preferences.favoriteColors !== undefined ? preferences.favoriteColors.join(' and ') : 'No favorite colors!';
}
",
                    "type": "ES_2020",
                  },
                  "dependencies": Array [
                    "preferences",
                    "undefined",
                    "preferences",
                  ],
                  "name": "formula_4b8f7e95470eda90057c4648aac4e4c7abb3f93559ed348246b6a15ec1fea473",
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Image",
                  "destination": "https://www.litmus.com/wp-content/uploads/2021/02/motion-tween-example.gif",
                  "nodes": Array [],
                  "title": "",
                },
              ],
            },
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.ListBlockDefinition",
              "delimiter": "period",
              "name": "favoriteColors",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.List",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Item",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                          "nodes": Array [
                            Object {
                              "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
                              "elementType": "String",
                              "name": "this",
                            },
                          ],
                        },
                      ],
                    },
                  ],
                  "tight": "true",
                  "type": "bullet",
                },
              ],
              "start": "1",
              "tight": "true",
              "type": "ordered",
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.ConditionalDefinition",
                  "condition": Object {
                    "$class": "org.accordproject.templatemark@0.4.0.Code",
                    "contents": "return preferences.favoriteColors !== undefined && preferences.favoriteColors.includes('PINK')",
                    "type": "ES_2020",
                  },
                  "dependencies": Array [
                    "preferences",
                    "undefined",
                    "preferences",
                  ],
                  "functionName": "condition_nodes_0_nodes_13_nodes_4_nodes_0",
                  "name": "favoriteColors",
                  "whenFalse": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Text",
                      "text": "Why don't you like PINK!",
                    },
                  ],
                  "whenTrue": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Text",
                      "text": "You like pink!",
                    },
                  ],
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Dynamic Query of Clauses",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "3",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Onboarding Clauses",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.FormulaDefinition",
                  "code": Object {
                    "$class": "org.accordproject.templatemark@0.4.0.Code",
                    "contents": "export function formula_a8a1d7714d95baa82f730e0105d2f2d0a9fc25ee9b6055058cc09667c01c01ab(data, library) {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    const now = dayjs();
    return jp.query(library, \`$.clauses[?(@.category==\\"onboarding\\")]\`);
}
",
                    "type": "ES_2020",
                  },
                  "dependencies": Array [
                    "jp",
                    "library",
                  ],
                  "name": "formula_a8a1d7714d95baa82f730e0105d2f2d0a9fc25ee9b6055058cc09667c01c01ab",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "3",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Authored by ",
            },
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
              "elementType": "String",
              "name": "firstName",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.FormulaDefinition",
                  "code": Object {
                    "$class": "org.accordproject.templatemark@0.4.0.Code",
                    "contents": "export function formula_daca9cb2f5bc16b65f544e6f408c1e3121d50a3251ec4fbe2f27132818acc3d2(data, library) {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    const now = dayjs();
    return jp.query(library, \`$.clauses[?(@.author==\\"\${firstName}\\")]\`);
}
",
                    "type": "ES_2020",
                  },
                  "dependencies": Array [
                    "jp",
                    "library",
                    "firstName",
                  ],
                  "name": "formula_daca9cb2f5bc16b65f544e6f408c1e3121d50a3251ec4fbe2f27132818acc3d2",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "3",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "High Risk",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.FormulaDefinition",
                  "code": Object {
                    "$class": "org.accordproject.templatemark@0.4.0.Code",
                    "contents": "export function formula_4c874b2977b5eef204e4987efc0a5594c7cce66403e54c17daa1e1c721d755a0(data, library) {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    const now = dayjs();
    return jp.query(library, \`$.clauses[?(@.risk>4)]\`);
}
",
                    "type": "ES_2020",
                  },
                  "dependencies": Array [
                    "jp",
                    "library",
                  ],
                  "name": "formula_4c874b2977b5eef204e4987efc0a5594c7cce66403e54c17daa1e1c721d755a0",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "3",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Low Risk and authored by ",
            },
            Object {
              "$class": "org.accordproject.templatemark@0.4.0.VariableDefinition",
              "elementType": "String",
              "name": "firstName",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.templatemark@0.4.0.FormulaDefinition",
                  "code": Object {
                    "$class": "org.accordproject.templatemark@0.4.0.Code",
                    "contents": "export function formula_646a6cadec2125e4fd9e4b756aac72bc618b528967f04d325b28a817774441dd(data, library) {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    const now = dayjs();
    return jp.query(library, \`$.clauses[?(@.risk<3 && @.author==\\"\${firstName}\\")]\`);
}
",
                    "type": "ES_2020",
                  },
                  "dependencies": Array [
                    "jp",
                    "library",
                    "firstName",
                  ],
                  "name": "formula_646a6cadec2125e4fd9e4b756aac72bc618b528967f04d325b28a817774441dd",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Done.",
            },
          ],
        },
      ],
    },
  ],
  "xmlns": "http://commonmark.org/xml/1.0",
}
`;
